File: 2. Array and String/19. unique_email_addresses/unique_email_addresses.js
/**
 * @param {string[]} emails
 * @return {number}
 */

// Using in-built functions
 var numUniqueEmails = function(emails) {
    let emailSet = new Set();
    for(let email of emails){
        
        let [localName, domainName] = email.split('@');
        localName = localName.split('+')[0].replaceAll('.','');
        emailSet.add(`${localName}@${domainName}`);
    }
    return emailSet.size;
};

// Without using inbuilt functions
// var numUniqueEmails = function(emails) {
//     let emailSet = new Set();
//     for(let email of emails){
//         let localName = '';
//         let domainName = '';
//         let index = 0;
//         while(email[index] !== '+' && email[index] !== '@'){
//             if(email[index] === '.') {
//                 index++;
//                 continue;
//             };
//             localName += email[index];
//             index++;
//         }
//         index=email.length - 1;
//         while(email[index] !== '@'){
//             domainName += email[index];
//             index--;
//         }
//         emailSet.add(`${localName}@${domainName}`);
//     }
//     return emailSet.size;
// }

// let ans = numUniqueEmails(["test.email+alex@leetcode.com","test.e.mail+bob.cathy@leetcode.com","testemail+david@lee.tcode.com"]);
// console.log(`Ans: ${ans}`);

// Complexity:
// Time Complexity: O(N.M)
// Space Complexity: O(N.M)

File: 2. Array and String/19. unique_email_addresses/unique_email_addresses.py

class Solution:
    def numUniqueEmails(self, emails):
        for index,email in enumerate(emails):
            emailCompList = email.split('@')
            localName = emailCompList[0]
            localName = localName.replace('.','').split('+')[0]
            emails[index] = f"{localName}@{emailCompList[1]}"
        
        return len(set(emails))

test = Solution()
print(test.numUniqueEmails(["test.email+alex@leetcode.com","test.e.mail+bob.cathy@leetcode.com","testemail+david@lee.tcode.com"]))
print(test.numUniqueEmails(["a@leetcode.com","b@leetcode.com","c@leetcode.com"]))
print(test.numUniqueEmails(["test.email+alex@leetcode.com","test.email.leet+alex@code.com"]))